### PR TITLE
Preparedness tab in region

### PR DESCRIPTION
**Pull request by shreeyash07**
## Addresses:
- [Preparedness- region](https://github.com/orgs/toggle-corp/projects/8/views/21?pane=issue&itemId=29916769)

## Changes
- Add HtmlOutput component for snippet response

## This PR doesn't introduce:
- [x] typos
- [x] conflict markers
- [x] unwanted comments
- [x] temporary files, auto-generated files or secret keys
- [x] `console.log` meant for debugging
- [x] codegen errors

### CODE REVIEW

File: src/views/Region/index.tsx [hunk 1]
```diff
@@ -149,14 +149,16 @@ export function Component() {
...
+                {(regionResponse && regionResponse?.preparedness_snippets.length > 0) ? (
+                    <NavigationTab
+                        to={generatePath(
+                            regionPreparednessRoute.absolutePath,
+                            { regionId },
+                        )}
+                    >
+                        {strings.regionPreparednessTab}
+                    </NavigationTab>
+                ) : null}
```
Comment (frozenhelium) [inline src/views/Region/index.tsx]: - Don't have to optional chain `regionResponse` when it's already checked
- Also check if `preparedness_snippets` is defined before accessing the length
- Even better if a variable is defined for this (e.g. `hasPreparedSnippets`)
- ? : can be replaced by &&
```suggestion
                {(regionResponse && regionResponse.preparedness_snippets && preparedness_snippets.length > 0) && (
                    <NavigationTab
                        to={generatePath(
                            regionPreparednessRoute.absolutePath,
                            { regionId },
                        )}
                    >
                        {strings.regionPreparednessTab}
                    </NavigationTab>
                ) }
```

File: src/views/RegionPreparedness/index.tsx [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
+                    <Container
+                        key={snippet.id}
+                        heading={snippet.title}
+                        withHeaderBorder
+                    >
+                        <HtmlOutput
+                            value={snippet.snippet}
+                        />
+                    </Container>
+                )
```
Comment (frozenhelium) [inline src/views/RegionPreparedness/index.tsx]: Missing trailing comma

**Review by frozenhelium — CHANGES_REQUESTED** [review]

File: src/utils/outletContext.ts [hunk 3]
```diff
@@ -11,8 +11,10 @@ export interface CountryOutletContext {
 }
 
 type RegionResponse = paths['/api/v2/region/{id}/']['get']['responses']['200']['content']['application/json'];
+type RegionKeyFigureResponse = paths['/api/v2/region_key_figure/']['get']['responses']['200']['content']['application/json'];
```
Comment (tnagorra) [inline src/utils/outletContext.ts]: Let's use helper GoApiResponse from restRequest

File: src/views/Region/index.tsx [hunk 4]
```diff
@@ -61,11 +72,24 @@ export function Component() {
...
         }),
-        [regionResponse],
+        [regionResponse, regionKeyFigureResponse],
     );
 
-    const pending = regionPending || aggregatedAppealPending;
+    const pending = regionPending || aggregatedAppealPending || regionKeyFigurePending;
+    const additionalInfoTabName = regionResponse?.additional_tab_name
+        ? regionResponse.additional_tab_name
+        : strings.regionAdditionalInfoTab;
```
Comment (tnagorra) [inline src/views/Region/index.tsx]: We can just use `??` or `||` to provide default value

File: src/views/RegionPreparedness/index.tsx [hunk 5]
```diff
@@ -1,12 +1,29 @@
...
+                        key={snippet.id}
+                        heading={snippet.title}
+                        withHeaderBorder
+                    >
+                        <HtmlOutput
+                            value={snippet.snippet}
+                        />
+                    </Container>
+                ),
+            )}
```
Comment (tnagorra) [inline src/views/RegionPreparedness/index.tsx]: Let's also handle case when the preparedness_snippes is empty.

@AdityaKhatri @frozenhelium Should we use List component instead? Or just use message component?

File: src/views/RegionAdditionalInfo/index.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                        >
+                            <TextOutput
+                                value={figure.source}
+                                label={strings.keyFigureLabel}
+                            />
+                        </KeyFigure>
+                    ))}
+                </Container>
+            )}
+            {hasRegionResponse && (
```
Comment (tnagorra) [inline src/views/RegionAdditionalInfo/index.tsx]: Should we show a message if we do not have data?

cc: @frozenhelium @AdityaKhatri

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/Region/index.tsx [hunk 7]
```diff
@@ -48,6 +49,16 @@ export function Component() {
         },
     });
 
+    const {
+        pending: regionKeyFigurePending,
+        response: regionKeyFigureResponse,
+    } = useRequest({
+        skip: !regionId,
```
Comment (tnagorra) [inline src/views/Region/index.tsx]:
```suggestion
        skip: isNotDefined(regionId),
```

File: src/views/Region/index.tsx [hunk 8]
```diff
@@ -61,11 +72,23 @@ export function Component() {
...
             regionResponse,
+            regionKeyFigureResponse,
         }),
-        [regionResponse],
+        [regionResponse, regionKeyFigureResponse],
     );
 
-    const pending = regionPending || aggregatedAppealPending;
+    const pending = regionPending || aggregatedAppealPending || regionKeyFigurePending;
+    const additionalInfoTabName = regionResponse?.additional_tab_name
```
Comment (tnagorra) [inline src/views/Region/index.tsx:80]:
```suggestion
    const additionalInfoTabName = !!regionResponse?.additional_tab_name
```

File: src/views/Region/index.tsx [hunk 9]
```diff
@@ -61,11 +72,23 @@ export function Component() {
...
+        || strings.regionAdditionalInfoTab;
+    const hasPreparednessSnippet = (
+        regionResponse
+        && regionResponse.preparedness_snippets.length > 0
+    );
+    const hasAdditionalInfoSnippet = (
+        regionResponse?.additional_tab_name
```
Comment (tnagorra) [inline src/views/Region/index.tsx]:
```suggestion
        !!regionResponse?.additional_tab_name
```

File: src/views/RegionPreparedness/index.tsx [hunk 10]
```diff
@@ -1,13 +1,63 @@
-import UnderConstructionMessage from '#components/UnderConstructionMessage';
+import { useCallback } from 'react';
+import { useOutletContext } from 'react-router-dom';
+import type { RegionOutletContext, RegionResponse } from '#utils/outletContext';
```
Comment (tnagorra) [inline src/views/RegionPreparedness/index.tsx]:
```suggestion
import { type RegionOutletContext, type RegionResponse } from '#utils/outletContext';
```

File: src/views/RegionAdditionalInfo/index.tsx [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        (_: number, data: RegionSnippetType): HtmlOutputProps => ({
+            className: styles.regionalContent,
+            value: data.snippet,
+        }),
+        [],
+    );
+
+    return (
+        <div className={styles.regionPreparedness}>
+            {hasKeyFigure && (
```
Comment (tnagorra) [inline src/views/RegionAdditionalInfo/index.tsx]: Let's handle case when hasKeyFigure and hasRegionResponse is false.

File: src/views/RegionAdditionalInfo/index.tsx [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    );
+}
+// eslint-disable-next-line import/prefer-default-export
+export function Component() {
+    const strings = useTranslation(i18n);
+    const { regionResponse, regionKeyFigureResponse } = useOutletContext<RegionOutletContext>();
+    const hasKeyFigure = (regionKeyFigureResponse?.results
+        && (regionKeyFigureResponse.results.length > 0));
+    const hasRegionResponse = regionResponse
+        && regionResponse?.snippets;
```
Comment (tnagorra) [inline src/views/RegionAdditionalInfo/index.tsx]: Let's also check length of the snippet

File: src/views/RegionAdditionalInfo/index.tsx [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        source,
+        id,
+        label,
+    } = props;
+
+    return (
+        <KeyFigure
+            // NOTE we should maintain key figures to take value as strings
+            // NOTE also design may vary from original we need to fix that too
+            key={id}
```
Comment (tnagorra) [inline src/views/RegionAdditionalInfo/index.tsx]: We don't need key here

File: src/views/RegionAdditionalInfo/index.tsx [hunk 14]
```diff
@@ -0,0 +1,122 @@
+import { useCallback } from 'react';
+import { useOutletContext } from 'react-router-dom';
+import type { RegionOutletContext, RegionKeyFigureResponse, RegionResponse } from '#utils/outletContext';
```
Comment (tnagorra) [inline src/views/RegionAdditionalInfo/index.tsx]: Let's change how type are imported.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/RegionPreparedness/i18n.json [hunk 15]
```diff
@@ -0,0 +1,6 @@
+{
+    "namespace": "regionPreparedness",
+    "strings": {
+        "NoDataMessage": "No Data to Display"
```
Comment (tnagorra) [inline src/views/RegionPreparedness/i18n.json]:
```suggestion
        "noDataMessage": "No data to display"
```

File: src/views/RegionPreparedness/index.tsx [hunk 16]
```diff
@@ -1,12 +1,77 @@
...
+    const { regionResponse } = useOutletContext<RegionOutletContext>();
+
+    const snippetListRendererParams = useCallback((_: number, data: RegionSnippet) => ({
+        id: data.id,
+        title: data.title,
+        snippet: data.snippet,
+    }), []);
+
+    const hasRegionResponse = (regionResponse
+        && (regionResponse?.snippets.length > 0));
```
Comment (tnagorra) [inline src/views/RegionPreparedness/index.tsx]: Why are we reading snippets when we are iterating over preparedness_snippets?

File: src/views/RegionPreparedness/index.tsx [hunk 17]
```diff
@@ -1,12 +1,77 @@
...
+                    renderer={Snippet}
+                    withMessage={false}
+                    pending={false}
+                    errored={false}
+                    filtered={false}
+                />
+            ) : (
+                <Message
+                    description={strings.NoDataMessage}
+                />
```
Comment (tnagorra) [inline src/views/RegionPreparedness/index.tsx]: This should be handled by the List component

File: src/views/RegionAdditionalInfo/i18n.json [hunk 18]
```diff
@@ -0,0 +1,9 @@
+{
+    "namespace": "regionAdditionalInfo",
+    "strings": {
+        "containerTitle": "Regional Content",
+        "keyFigureHeading": "Key Figure",
+        "keyFigureLabel": "Source",
+        "NoDataMessage": "No Data to Display"
```
Comment (tnagorra) [inline src/views/RegionAdditionalInfo/i18n.json]:
```suggestion
        "NoDataMessage": "No data to display"
```

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by tnagorra — APPROVED** [review]

File: src/components/List/index.tsx [hunk 19]
```diff
@@ -47,6 +49,8 @@ function List<DATUM, KEY extends ListKey, RENDERER_PROPS>(
         filteredMessage = 'Data is not available for the selected filter!',
 
         compact,
+
+        withMessage = true,
```
Comment (frozenhelium) [inline src/components/List/index.tsx]: We should use opposite prop and make it by default false

Comment (frozenhelium) [reply src/components/List/index.tsx]: We'll later on add a bare list to fix this

File: src/views/RegionAdditionalInfo/styles.module.css [hunk 20]
```diff
@@ -0,0 +1,32 @@
+.region-preparedness {
```
Comment (frozenhelium) [inline src/views/RegionAdditionalInfo/styles.module.css]: This should match the name of component
```suggestion
.region-additional-info {
```

File: src/views/RegionAdditionalInfo/styles.module.css [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+            justify-content: center;
+            gap: var(--go-ui-spacing-2xl);
+
+            .region-key-figure {
+                align-items: center;
+                flex-basis: 20rem;
+                border-radius: var(--go-ui-border-radius-md);
+                box-shadow: var(--go-ui-box-shadow-lg);
+                height: 100%;
+            }
```
Comment (frozenhelium) [inline src/views/RegionAdditionalInfo/styles.module.css]: We should use grid for the styling here

File: src/views/RegionAdditionalInfo/index.tsx [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+    const regionSnippetParams = useCallback(
+        (_: number, data: RegionSnippetType): HtmlOutputProps => ({
+            className: styles.regionalContent,
+            value: data.snippet,
+        }),
+        [],
+    );
+
+    return (
+        <div className={styles.regionPreparedness}>
```
Comment (frozenhelium) [inline src/views/RegionAdditionalInfo/index.tsx]: We should use component name for the main container styling
```suggestion
        <div className={styles.regionAdditionalInfo}>
```

**Review by frozenhelium — CHANGES_REQUESTED** [review]

**Review by frozenhelium — APPROVED** [review]

**Pull request merged**